src/RFC-0306_DANTemplateRegistration.md

```diff
@@ -0,0 +1,173 @@
+composed with other templates and run, by a set of layer 2 [validator node]s.
+
+This RFC is primarily concerned with the mechanism for making these templates available to other parties wishing to
+build contracts and run them. The most important aspect of this is to allow any [validator node] to verify that they
+are running the same code as other members of the same committee.
+
+## Template Registration UTXO fields
+
+A base-layer template registration [UTXO] MUST have the `TEMPLATE_REGISTRATION` output flag set and contain the following data:
+
+| Name              | Type               | Description                                                                                                                           |
+|-------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------|
+| author_public_key | ECC public key     | Public key of the author                                                                                                              |
+| author_signature  | Schnorr signature  | Signature that signs remaining fields                                                                                                 |
+| template_name     | String (255)       | A descriptive name for the template                                                                                                   |
+| template_version  | Varint             | Code version as a single number                                                                                                       |
+| build_info        | BuildInfo struct   | Information on the binary build                                                                                                       |                                                                                                                             |
+| binary_checksum   | SHA2 checksum      | A SHA2 checksum of the WASM binary.                                                                                                   |
+| binary_address    | [Multiaddr]        | A [multiaddr] pointing to the WASM module binary download. This may be an HTTP, ONION, p2p, or IPFS address. Maximum byte length: 255 |
+
+The `author_signature` is a Schnorr proof that commits to the template fields contained in the [UTXO], namely
+`template_name`, `template_version`, `build_info`, `binary_checksum`, and `binary_address`.
+
+The base node acts as a notary for this data, it is not responsible for the validity of the template fields. However, it must
+not allow malformed/invalid data to be committed to the blockchain.
+
+Therefore, some additional base-layer consensus rules are required for a `TEMPLATE_REGISTRATION` [UTXO]:
+* A base node MUST validate the `author_signature`; and
+* The [UTXO] MUST have a relative time-lock of 100 blocks.
```

Comment:
What's the reason for the lock height?
I can think of reasons why you might want to push an updated version shortly. For example, you realized there is a bug in the just pushed version, now you have to wait for 200 minutes before you can push your fix?

Reply:
Yeah this was chosen pretty arbitrarily. I think it's more around having some assurance for this UTXO existing on the blockchain when referencing this template in a contract - which this doesn't really answer so I'll remove this.

Will have to put more thought into how we reference template UTXOs or if we copy the template data into the contract definition/constitution which is kind of bloaty.